File: js/dialog_preferences.js
"use strict";

/**
 * App-wide dialog suppression prefs: ~/.masonjar/dialog_preferences.json
 * Schema shared with py/dialog_preferences.py
 */

var fs = require("fs");
var path = require("path");
var homeDir = require("./home_dir");

var FILENAME = "dialog_preferences.json";

var KEY_MIXED_RESOLUTION_TIER = "adjust.mixed_resolution_tier";
var KEY_CONFIRM_SAVE_OVERWRITE = "adjust.confirm_save_overwrite";
var KEY_ISOLATE_LABEL_AUDIT = "adjust.isolate_label_audit";

var KEY_LABELS = {};
KEY_LABELS[KEY_MIXED_RESOLUTION_TIER] =
	"Viewer/Editor: mixed-resolution tier change notice";
KEY_LABELS[KEY_CONFIRM_SAVE_OVERWRITE] =
	"Viewer/Editor: confirm before overwriting annotation on Save";
KEY_LABELS[KEY_ISOLATE_LABEL_AUDIT] =
	"Viewer/Editor: Isolate Regions notice after Save";

function prefsPath(homeRoot) {
	return path.join(homeDir.masonHomePath(homeRoot), FILENAME);
}

function defaultPrefs(appVersion) {
	return { app_version: String(appVersion || ""), suppressed: {} };
}

function load(homeRoot) {
	var filePath = prefsPath(homeRoot);
	if (!fs.existsSync(filePath)) {
		return defaultPrefs();
	}
	try {
		var raw = JSON.parse(fs.readFileSync(filePath, "utf8"));
		if (!raw || typeof raw !== "object") {
			return defaultPrefs();
		}
		var suppressed = raw.suppressed && typeof raw.suppressed === "object"
			? raw.suppressed
			: {};
		var out = { app_version: String(raw.app_version || ""), suppressed: {} };
		Object.keys(suppressed).forEach(function (k) {
			out.suppressed[k] = !!suppressed[k];
		});
		return out;
	} catch (_err) {
		return defaultPrefs();
	}
}

function save(prefs, homeRoot) {
	var filePath = prefsPath(homeRoot);
	fs.mkdirSync(path.dirname(filePath), { recursive: true });
	var payload = {
		app_version: String((prefs && prefs.app_version) || ""),
		suppressed: (prefs && prefs.suppressed) || {},
	};
	fs.writeFileSync(filePath, JSON.stringify(payload, null, 2) + "\n", "utf8");
}

function clearSuppressions(homeRoot, appVersion) {
	var prefs = load(homeRoot);
	prefs.suppressed = {};
	if (appVersion != null) {
		prefs.app_version = String(appVersion);
	}
	save(prefs, homeRoot);
	return prefs;
}

function syncAppVersionClearIfChanged(currentVersion, homeRoot) {
	var prefs = load(homeRoot);
	var stored = String(prefs.app_version || "");
	var cur = String(currentVersion || "");
	if (stored === cur && cur) {
		return false;
	}
	prefs.suppressed = {};
	prefs.app_version = cur;
	save(prefs, homeRoot);
	return true;
}

function listSuppressed(homeRoot) {
	var prefs = load(homeRoot);
	var suppressed = prefs.suppressed || {};
	return Object.keys(suppressed)
		.filter(function (k) {
			return !!suppressed[k];
		})
		.map(function (k) {
			return { key: k, label: KEY_LABELS[k] || k };
		});
}

module.exports = {
	FILENAME: FILENAME,
	KEY_MIXED_RESOLUTION_TIER: KEY_MIXED_RESOLUTION_TIER,
	KEY_CONFIRM_SAVE_OVERWRITE: KEY_CONFIRM_SAVE_OVERWRITE,
	KEY_ISOLATE_LABEL_AUDIT: KEY_ISOLATE_LABEL_AUDIT,
	KEY_LABELS: KEY_LABELS,
	prefsPath: prefsPath,
	load: load,
	save: save,
	clearSuppressions: clearSuppressions,
	syncAppVersionClearIfChanged: syncAppVersionClearIfChanged,
	listSuppressed: listSuppressed,
};

File: py/dialog_preferences.py
"""App-wide dialog suppression preferences under ~/.masonjar/dialog_preferences.json."""

from __future__ import annotations

import json
from pathlib import Path
from typing import Any

DIALOG_PREFS_FILENAME = "dialog_preferences.json"

KEY_MIXED_RESOLUTION_TIER = "adjust.mixed_resolution_tier"
KEY_CONFIRM_SAVE_OVERWRITE = "adjust.confirm_save_overwrite"
KEY_ISOLATE_LABEL_AUDIT = "adjust.isolate_label_audit"

KNOWN_KEYS = (
    KEY_MIXED_RESOLUTION_TIER,
    KEY_CONFIRM_SAVE_OVERWRITE,
    KEY_ISOLATE_LABEL_AUDIT,
)


def mason_home_dir() -> Path:
    return Path.home() / ".masonjar"


def dialog_preferences_path(home: Path | None = None) -> Path:
    return (home or mason_home_dir()) / DIALOG_PREFS_FILENAME


def default_prefs(app_version: str = "") -> dict[str, Any]:
    return {"app_version": str(app_version or ""), "suppressed": {}}


def load_dialog_preferences(home: Path | None = None) -> dict[str, Any]:
    path = dialog_preferences_path(home)
    if not path.is_file():
        return default_prefs()
    try:
        raw = json.loads(path.read_text(encoding="utf-8"))
    except (OSError, json.JSONDecodeError, TypeError):
        return default_prefs()
    if not isinstance(raw, dict):
        return default_prefs()
    suppressed = raw.get("suppressed")
    if not isinstance(suppressed, dict):
        suppressed = {}
    return {
        "app_version": str(raw.get("app_version") or ""),
        "suppressed": {str(k): bool(v) for k, v in suppressed.items()},
    }


def save_dialog_preferences(prefs: dict[str, Any], home: Path | None = None) -> None:
    path = dialog_preferences_path(home)
    path.parent.mkdir(parents=True, exist_ok=True)
    payload = {
        "app_version": str(prefs.get("app_version") or ""),
        "suppressed": dict(prefs.get("suppressed") or {}),
    }
    path.write_text(json.dumps(payload, indent=2) + "\n", encoding="utf-8")


def is_suppressed(key: str, home: Path | None = None) -> bool:
    prefs = load_dialog_preferences(home)
    return bool((prefs.get("suppressed") or {}).get(key))


def set_suppressed(key: str, value: bool = True, home: Path | None = None) -> None:
    prefs = load_dialog_preferences(home)
    suppressed = dict(prefs.get("suppressed") or {})
    if value:
        suppressed[key] = True
    else:
        suppressed.pop(key, None)
    prefs["suppressed"] = suppressed
    save_dialog_preferences(prefs, home)


def clear_suppressions(home: Path | None = None, app_version: str | None = None) -> dict[str, Any]:
    prefs = load_dialog_preferences(home)
    prefs["suppressed"] = {}
    if app_version is not None:
        prefs["app_version"] = str(app_version)
    save_dialog_preferences(prefs, home)
    return prefs


def sync_app_version_clear_if_changed(
    current_version: str, home: Path | None = None
) -> bool:
    """If stored app_version differs from current, clear suppressions. Returns True if cleared."""
    prefs = load_dialog_preferences(home)
    stored = str(prefs.get("app_version") or "")
    cur = str(current_version or "")
    if stored == cur and cur:
        return False
    prefs["suppressed"] = {}
    prefs["app_version"] = cur
    save_dialog_preferences(prefs, home)
    return True
